### PR TITLE
feature: optimise the logic of using prometheus

### PR DESCRIPTION
**Pull request by fengzixu**
Signed-off-by:  fengzixu <hnustphoenix@gmail.com>

### Ⅰ. Describe what this PR did

1. use the custom registry to replace the default.
2. upgrade the package called client-go of prometheus
3. add the new dependent package called promhttp
4. enable metrics about gRPC server

Related to:  https://github.com/alibaba/pouch/issues/2395

### Ⅱ. Does this pull request fix one issue?

#2395 #2414 

### Ⅳ. Describe how to verify it

First, I complied a new binary pouchd, and I run it with parameters: 
```
 /home/xr/workspace/go-project/src/github.com/alibaba/pouch/bin/pouchd -l tcp://0.0.0.0:4243 -l unix:///var/run/pouchd.sock --enable-cri=true --cri-version=v1alpha2
```
Second, I visted successfully the url: `http://<hostIP>:4243/metrics` so that I can capture metrics produced by pouchd.

image

And then, I check some points about metrics:

#### Is there all metrics was registered successfully?

I saved the result about metrics: https://gist.github.com/fengzixu/ab018bd1d1fa8ddc56a6fe9150d36feb. I picked up some samples and compared with original metrics in code. There is no any difference.

#### Are the active metrics show commonly after I process some operations ?

I use the command `pouch pull nginx` to pull a image, and I re-captured the metrics from pouch. The figure of metrics about image has a little increase.

#### Whether the metrics about gRPC server show commonly?

I started up a k8s cluster together with pouch. And I created some pods. After that I found that the metrics about gRPC between pouchd and kubelet has some figures. It was beneficial for me when I debug this function.

image

### CODE REVIEW

**Comment by fengzixu** [comment]
@Starnop After I developed this PR and thought deeply about metrics of gRPC server. I considered that we should save the gRPC metrics. That was beneficial for developers when they want to develop some functions,  which perhaps were associated with communication between kubelet and pouchd.

**Comment by fengzixu** [comment]
Request review.

**Comment by fengzixu** [comment]
By the way, the main reason I upgrade the package called `github.com/prometheus/client_golang/prometheus` is that I want to use package called `promehttp`.

File: vendor/vendor.json [hunk 1]
```diff
@@ -1568,10 +1568,16 @@
...
+			"checksumSHA1": "WSCsSdD3I7Are0ntaIq6zR9tl0s=",
 			"path": "github.com/prometheus/client_golang/prometheus",
-			"revision": "5cec1d0429b02e4323e042eb04dafdb079ddf568",
-			"revisionTime": "2017-10-05T11:29:15Z"
+			"revision": "f02bfc3484a6b03d1fc00d72d86add103ef9567b",
+			"revisionTime": "2018-01-11T15:59:30Z"
+		},
+		{
+			"checksumSHA1": "wJzzub/0w2espYvar3lylwHXBAk=",
+			"path": "github.com/prometheus/client_golang/prometheus/promhttp",
```
Comment (allencloud) [inline vendor/vendor.json]: If we are using an inner package of `github.com/prometheus/client_golang/prometheus`.
I wish that we use the same revision of of the package.

But here you have different reversion of `f02bfc3484a6b03d1fc00d72d86add103ef9567b ` and `f30f428035633da15d00d3dfefb0128c5e569ef4 `. WDYT?

Comment (fengzixu) [reply vendor/vendor.json]: I use the package `promhttp` whose version is same with `github.com/prometheus/client_golang/prometheus`

**Comment by fuweid** [comment]
> By the way, the main reason I upgrade the package called github.com/prometheus/client_golang/prometheus is that I want to use package called promehttp.

Sorry, I din't get the point about this PR. Could you explain more detail about what is good from using customize registry? Thanks

**Comment by fengzixu** [comment]
> > By the way, the main reason I upgrade the package called github.com/prometheus/client_golang/prometheus is that I want to use package called promehttp.
> 
> Sorry, I din't get the point about this PR. Could you explain more detail about what is good from using customize registry? Thanks

I describe some details about this PR in  https://github.com/alibaba/pouch/issues/2395. @fuweid

**Comment by fuweid** [comment]
> According to those comments about DefaultRegisterer, we can know that there is a big risk about using global variant in our code especially in the dependent package. If someone modifies this variant accidentally through some methods written in prometheus. That is a disaster. That is a reason, we should avoid using global variants as much as possible.

Good to know. Thanks

File: vendor/vendor.json [hunk 2]
```diff
@@ -1930,4 +1936,4 @@
 		}
 	],
 	"rootPath": "github.com/alibaba/pouch"
-}
+}
```
Comment (Starnop) [inline vendor/vendor.json]: There should not be a change, please handle it, THX.

**Comment by Starnop** [comment]
Except for the little problem above, LGTM. Thanks for your contribution. :)

File: apis/server/router.go [hunk 3]
```diff
@@ -88,8 +89,8 @@ func initRoute(s *Server) http.Handler {
 	s.addRoute(r, http.MethodPost, "/networks/{id:.*}/disconnect", s.disconnectNetwork)
 
 	// metrics
-	r.Path(versionMatcher + "/metrics").Methods(http.MethodGet).Handler(prometheus.Handler())
-	r.Path("/metrics").Methods(http.MethodGet).Handler(prometheus.Handler())
+	r.Path(versionMatcher + "/metrics").Methods(http.MethodGet).Handler(promhttp.HandlerFor(util_metrics.GetCustomPrometheusRegistry(), promhttp.HandlerOpts{}))
```
Comment (HusterWan) [inline apis/server/router.go]: How to make the function length short?

Comment (fengzixu) [reply apis/server/router.go]: Ok. Let me fix it.

**Comment by fengzixu** [comment]
@HusterWan @Starnop Request review again. I have fixed some problems above.

**Comment by fengzixu** [comment]
After I finished this PR, I have some throughts about using prometheus:

With the development of pouch,  we perhaps have more and more requirements about monitoring, especially prometheus and gRPC. Whether we shoud abstract and encapsulate a common package which can support some universal and standard approaches about add monitor to developers especially don't familar with Prometheus. @allencloud @Starnop @fuweid

File: pkg/utils/metrics/metrics.go [hunk 4]
```diff
@@ -63,3 +80,21 @@ func NewLabelTimer(subsystem, name, help string, labels ...string) *prometheus.H
 			ConstLabels: nil,
 		}, labels)
 }
+
+// GetCustomPrometheusRegistry create a custom resigtry of Prometheus.
+func GetCustomPrometheusRegistry() *prometheus.Registry {
```
Comment (fuweid) [inline pkg/utils/metrics/metrics.go]: I think we should use `GetPrometheusRegistry` here. The `Custom` is for the prometheus package, not for pouch. Right?

**Comment by HusterWan** [comment]
lgtm

**Comment by allencloud** [comment]
Thanks for your contribution. @fengzixu

**Comment by fuweid** [comment]
@fengzixu could you follow up to change the name? thanks

**Pull request merged**